File: java/leetcode/MaximizeDistance.java
//Leetcode problem 849 Maximize Distance to Closest Person
//Solution written by Xuqiang Fang on 10 June, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
class Solution{
    public int maxDistToClosest(int[] seats) {
        int n = seats.length;
        int l = 0;
        int r = 1;
        int m = 1;
        for(int i=1; i<n; ++i){
            if(seats[i] == 1){
                r = i;
                if(seats[0] == 0 && l == 0){
                    m = Math.max(m, 2*(r-l));
                }
                else{
                    m = Math.max(m, r-l);
                }
                l = r;
            }
            else if(i == n-1){
                m = Math.max(m, 2*(n-1-l));
            }
        }
        return m/2;
    }
}

public class MaximizeDistance{
	public static void main(String[] args){
		Solution s = new Solution();
        int[] seats = {1,0,0,0,1,0,1};
        int[] seat = {1,0,0,0};
        int[] sea = {0,0,0,0,0,1};
        System.out.println(s.maxDistToClosest(seats));
        System.out.println(s.maxDistToClosest(seat));
        System.out.println(s.maxDistToClosest(sea));
	}
}

File: cpp/leetcode/MaximizeDistance.cpp
//Leetcode Problem No 849 Maximize Distance to Closest Person
//Solution written by Xuqiang Fang on 10 June, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class Solution{
public:
    int maxDistToClosest(vector<int>& s) {
        const int n = s.size();
        int m = 1;
        int l = 0, r = 0;
        for(int i=1; i<n; ++i){
            if(s[i] == 1){
                r = i;
                if(s[0] == 0 && l == 0){
                    m = max(m, 2*(r-l));
                }
                else{
                    m = max(m, r-l);
                }
                l = r;
            }
            else if(i == n-1){
                m = max(m, 2*(n-1-l));
            }
        }
        return m/2;
    }
};

int main(){
    Solution s;
    vector<int> seats{1,0,0,0,1,0,1};
    vector<int> seat{1,0,0,0};
    vector<int> sea{1,0};
    vector<int> se{0,0,0,0,0,1};
    cout << s.maxDistToClosest(seats) << endl;
    cout << s.maxDistToClosest(seat) << endl;
    cout << s.maxDistToClosest(sea) << endl;
    cout << s.maxDistToClosest(se) << endl;
    return 0;
}
